enhancements/20240709-longhorn-v2-data-engine-v1.4.0.md

```diff
@@ -0,0 +1,174 @@
+    - When a given NodeConfig changes to enable the V2 engine:
+        - Create `/oem/93_spdk.yaml`
+        - Run `modprobe vfio_pci ; modprobe uio_pci_generic ; modprobe nvme_tcp`
+        - Run `echo 1024 > /sys/kernel/mm/hugepages/hugepages-2048kB/nr_hugepages`
+        - If any of the modprobes fail, or the hugepages can't be allocated, record this somewhere so that it can be picked up in the Harvester GUI with a suggestion to the user to reboot the node (TODO: figure out how to actually do this).
+    - If the V2 engine becomes disabled:
+        - Remove `/oem/93_spdk.yaml`
+        - Run `modprobe -r vfio_pci ; modprobe -r uio_pci_generic ; modprobe -r nvme_tcp`
+        - Run `echo 0 > /sys/kernel/mm/hugepages/hugepages-2048kB/nr_hugepages`
+
+[os2#98]: https://github.com/harvester/os2/pull/98
+[Longhorn Quick Start]: https://longhorn.io/docs/1.6.2/v2-data-engine/quick-start/
+
+#### Adding Disks ####
+
+This is largely covered by Vicente's [node-disk-manager refactor HEP], which proposes adding a Longhorn V2 provisioner to NDM.
+
+Note that we will also need to teach NDM to ignore V2 volumes attached to the harvester host, much as it currently ignores V1 volumes by looking for the string "longhorn" in the block device's bus path.  For V2 volumes, we end up with a bunch of `/dev/nvme*` and `/dev/dm-*` devices.  We should probably just ignore the latter by device path.  For nvme devices, here's an example of the udev data, which suggests we can look for "SPDK_bdev_Controller" in a variety of fields:
+
+```
+# cat /var/run/udev/data/b259:2
+S:disk/by-id/nvme-uuid.c67ce251-57bc-4dd4-b2a4-60cccbc5f63b
+S:disk/by-id/nvme-SPDK_bdev_Controller_00000000000000000000_1
+S:disk/by-path/nvme-1
+S:disk/by-id/nvme-SPDK_bdev_Controller_00000000000000000000
+I:1713513798
+E:MPATH_SBIN_PATH=/sbin
+E:DM_MULTIPATH_DEVICE_PATH=0
+E:ID_SERIAL_SHORT=00000000000000000000
+E:ID_WWN=uuid.c67ce251-57bc-4dd4-b2a4-60cccbc5f63b
+E:ID_MODEL=SPDK bdev Controller
+E:ID_REVISION=23.05
+E:ID_NSID=1
+E:ID_SERIAL=SPDK_bdev_Controller_00000000000000000000_1
+E:ID_PATH=nvme-1
+E:ID_PATH_TAG=nvme-1
+E:COMPAT_SYMLINK_GENERATION=2
+E:ID_FS_TYPE=
+G:systemd
+Q:systemd
+V:1
+```
+
+[node-disk-manager refactor HEP]: https://github.com/harvester/harvester/pull/6015
+
+#### Creating Storage Classes
+
+We need to add the "dataEngine" parameter to the Harvester GUI, so that when creating a Longhorn storage class, the user can select either V1 or V2 Data Engine.  If V2 Data Engine is selected, Migratable must be automatically set to "No", and the Migratable field disabled so the user can't change it (the V2 data engine does not yet support live migration).
+
+#### Creating Volumes #####
+
+The Harvester GUI needs to detect when a Longhorn V2 storage class is selected during volume creation, and use this as a trigger for setting `accessMode: ReadWriteOnce` on the PVC that will be created (again, this is because the V2 data engine does not yet support live migration).
```

Comment:
When the Longhorn v2 volume supports live migration, should we add `Migratable: True` and `accessMode: ReadWriteMany` back like the v1 volume?

NOTE: the v1 volume depends on the `Migratable: True` to determine whether the `block` mode volume or `filesystem` mode volume.

Reply:
> When the Longhorn v2 volume supports live migration, should we add `Migratable: True` and `accessMode: ReadWriteMany` back like the v1 volume?

Yeah, I think so.